### PR TITLE
[Tennistv] Rewrite

### PR DESCRIPTION
**Pull request by zenerdi0de**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Rewrite cause site got updated and no longer uses the older API endpoints.

### CODE REVIEW

File: yt_dlp/extractor/tennistv.py [hunk 1]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+                'grant_type': 'authorization_code',
+                'client_id': 'tennis-tv-web',
+                'redirect_uri': 'https://www.tennistv.com/resources/v1.1.10/html/silent-check-sso.html'
+            })
+            return
+
+        if cookies is not None:
+            self.ACCESS_TOKEN = cookies.get('access_token').value
+            self.REFRESH_TOKEN = cookies.get('refresh_token').value
+            return
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:96]: make non-fatal

File: yt_dlp/extractor/tennistv.py [hunk 2]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+
+        else:
+            self.report_warning('Site requires a sbuscribed account. Either pass in the cookies or login credentials')
+
+    def token_util(self, video_id, payload):
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:101]: token_util? did u mean get_token?

File: yt_dlp/extractor/tennistv.py [hunk 3]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+                'password': password,
+                'submitAction': 'Log In'
+            }
+
+            login_page = self._download_webpage(update_url_query(BASE_URL, {'client_id': 'tennis-tv-web',
+                                                'redirect_uri': 'https://tennistv.com',
+                                                                            'response_mode': 'fragment',
+                                                                            'response_type': 'code',
+                                                                            'scope': 'openid'}),
+                                                '')
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:66]:
```suggestion
            login_page = self._download_webpage(BASE_URL, None, query={
                'client_id': 'tennis-tv-web',
                'redirect_uri': 'https://tennistv.com',
                'response_mode': 'fragment',
                'response_type': 'code',
                'scope': 'openid'
            }))
```
etc

File: yt_dlp/extractor/tennistv.py [hunk 4]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+            post_url = self._html_search_regex(r'action=["\'](.+?)["\']\s+?method=["\']post["\']', login_page, None)
+
+            temp_page = self._download_webpage(post_url, None, 'Sending login data', 'Unable to send login data',
+                                               data=urlencode_postdata(login_form), headers=self.headers)
+
+            if 'Your username or password was incorrect' in temp_page:
+                raise ExtractorError('Your username or password was incorrect', expected=True)
+
+            _, handle = self._download_webpage_handle(update_url_query(BASE_URL, {'client_id': 'tennis-tv-web',
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:75]: _request_webpage

File: yt_dlp/extractor/tennistv.py [hunk 5]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+        return self._download_json(f'https://atppayments.streamamg.com/api/v1/session/ksession/?lang=en&apijwttoken={self.ACCESS_TOKEN}&entryId={entryid}',
+                                   video_id, 'Downloading ksession token', 'Failed to download ksession token', headers=self.headers)
+
     def _real_extract(self, url):
         video_id = self._match_id(url)
         webpage = self._download_webpage(url, video_id)
 
-        internal_id = self._search_regex(r'video=([\w-]+)', webpage, 'internal video id')
+        title = self._html_search_regex(r'<title>(.+?)</title>', webpage, 'Title',
+                                        fatal=False) or self._og_search_title(webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:121]:
```suggestion
        title = self._html_search_regex((r'<title>(.+?)</title>', *self._og_regexes('title')),
                                        webpage, 'Title', fatal=False)
```

File: yt_dlp/extractor/tennistv.py [hunk 6]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+            k_session = self._download_session_json(video_id, entryid).get('KSession')
 
-        series = vdata['tour']
-        venue = vdata['displayText']['venue']
-        round_str = vdata['seo']['round']
+            if k_session is None:
+                raise ExtractorError('Failed to get KSession, possibly a premium video', expected=True)
+
+        if session_json.get('ErrorMessage'):
+            self.to_screen(session_json['ErrorMessage'])
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:142]: report_warning?

File: yt_dlp/extractor/tennistv.py [hunk 7]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(f'https://open.http.mp.streamamg.com/p/{self.PARTNER_ID}/sp/{self.PARTNER_ID}00/playManifest/entryId/{entryid}/format/applehttp/protocol/https/a.m3u8?ks={k_session}', video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:144]: too long. the URL can be seperated out
```py
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            self._FORMAT_URL.format(partner=self.PARTNER_ID, entry=entryid, session=k_session), video_id)
```

File: yt_dlp/extractor/tennistv.py [hunk 8]
```diff
@@ -31,84 +34,123 @@ class TennisTVIE(InfoExtractor):
...
+
+        self._sort_formats(formats)
 
         return {
             'id': video_id,
             'title': title,
-            'description': description,
+            'description': self._html_search_regex(r'<span itemprop="description" content=["\'](.+?)["\']>', webpage, 'description', fatal=False) or self._og_search_description(webpage),
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:151]: line too long

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tennistv.py [hunk 9]
```diff
@@ -1,13 +1,16 @@
 # coding: utf-8
 from __future__ import unicode_literals
 
-import json
-
 from .common import InfoExtractor
 
+from ..compat import compat_urlparse
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:6]:
```suggestion
import urllib.parse
```

File: yt_dlp/extractor/tennistv.py [hunk 10]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+        'content-Type': 'application/x-www-form-urlencoded'
+    }
 
-        login_result = self._download_json(
-            'https://www.tennistv.com/api/users/v1/login', None,
-            note='Logging in',
-            errnote='Login failed (wrong password?)',
-            headers=headers,
-            data=login_json)
+    def _login(self):
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:47]: See the new documentation on adding login support. The merge conflict should disappear once this is fixed
https://github.com/yt-dlp/yt-dlp/blob/ae6a1b95857c7ba1392e37be99b9ee2b66b14b1d/yt_dlp/extractor/common.py#L437-L443

File: yt_dlp/extractor/tennistv.py [hunk 11]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+            })
+            return
+
+        if cookies.get('access_token') and cookies.get('refresh_token') is not None:
+            self.ACCESS_TOKEN = cookies.get('access_token').value
+            self.REFRESH_TOKEN = cookies.get('refresh_token').value
+            return
+
+        else:
+            self.report_warning('Site requires a subscribed account. Either pass in the cookies or login credentials.')
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:102]: If all videos need subscription, better to `self.raise_login_required`

Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:102]:
```suggestion
        if cookies.get('access_token') and cookies.get('refresh_token'):
            self.ACCESS_TOKEN = cookies['access_token'].value
            self.REFRESH_TOKEN = cookies['refresh_token'].value
        else:
            self.raise_login_required()
```

File: yt_dlp/extractor/tennistv.py [hunk 12]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+            handle = self._request_webpage(update_url_query(BASE_URL, query={
+                'client_id': 'tennis-tv-web',
+                'redirect_uri': 'https://www.tennistv.com/resources/v1.1.10/html/silent-check-sso.html',
+                'state': random_uuidv4(),
+                'response_mode': 'fragment',
+                'response_type': 'code',
+                'scope': 'openid',
+                'nonce': random_uuidv4(),
+                'prompt': 'none'}),
+                '', headers=self.headers)
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:86]:
```py
            handle = self._request_webpage(BASE_URL, query=..., headers=self.headers)
```

File: yt_dlp/extractor/tennistv.py [hunk 13]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+
+            self.get_token(None, {
+                'code': compat_urlparse.parse_qs(handle.geturl()).get('code')[-1],  # There can be multiple, always use the last one
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:89]:
```suggestion
                'code': compat_urlparse.parse_qs(handle.geturl()).get('code')[-1],
```

File: yt_dlp/extractor/tennistv.py [hunk 14]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(self._FORMAT_URL.format(partner=self.PARTNER_ID, entry=entryid, session=k_session), video_id)
+
+        self._sort_formats(formats)
 
         return {
             'id': video_id,
             'title': title,
-            'description': description,
+            'description': self._html_search_regex((r'<span itemprop="description" content=["\'](.+?)["\']>', *self._og_regexes('description')), webpage, 'description', fatal=False),
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:154]:
```suggestion
            'description': self._html_search_regex(
                (r'<span itemprop="description" content=["\']([^"\']+)["\']>', *self._og_regexes('description')),
                webpage, 'description', fatal=False),
```

File: yt_dlp/extractor/tennistv.py [hunk 15]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
             'formats': formats,
-            'thumbnail': thumbnail,
-            'timestamp': timestamp,
-            'series': series,
-            'season': venue,
-            'episode': round_str,
+            'thumbnail': f'https://open.http.mp.streamamg.com/p/{self.PARTNER_ID}/sp/{self.PARTNER_ID}00/thumbnail/entry_id/{entryid}/version/100001/height/1920',
+            'timestamp': unified_timestamp(self._html_search_regex(r'<span itemprop="description" content=["\'](.+?)["\']>', webpage, 'upload time')),
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:157]:
```suggestion
            'timestamp': unified_timestamp(self._html_search_regex(
                r'<span itemprop="description" content=["\']([^"\']+)["\']>', webpage, 'upload time')),
```

File: yt_dlp/extractor/tennistv.py [hunk 16]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
-        series = vdata['tour']
-        venue = vdata['displayText']['venue']
-        round_str = vdata['seo']['round']
+            if k_session is None:
+                raise ExtractorError('Failed to get KSession, possibly a premium video', expected=True)
+
+        if session_json.get('ErrorMessage'):
+            self.report_warning(session_json['ErrorMessage'])
+
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(self._FORMAT_URL.format(partner=self.PARTNER_ID, entry=entryid, session=k_session), video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:147]:
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            self._FORMAT_URL.format(partner=self.PARTNER_ID, entry=entryid, session=k_session), video_id)
```

File: yt_dlp/extractor/tennistv.py [hunk 17]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+
+        self.ACCESS_TOKEN = res.get('access_token')
+        self.REFRESH_TOKEN = res.get('refresh_token')
 
     def _real_initialize(self):
         self._login()
 
+    def _download_session_json(self, video_id, entryid,):
+        return self._download_json(f'https://atppayments.streamamg.com/api/v1/session/ksession/?lang=en&apijwttoken={self.ACCESS_TOKEN}&entryId={entryid}',
+                                   video_id, 'Downloading ksession token', 'Failed to download ksession token', headers=self.headers)
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:117]:
```suggestion
        return self._download_json(
            f'https://atppayments.streamamg.com/api/v1/session/ksession/?lang=en&apijwttoken={self.ACCESS_TOKEN}&entryId={entryid}',
            video_id, 'Downloading ksession token', 'Failed to download ksession token', headers=self.headers)
```

File: yt_dlp/extractor/tennistv.py [hunk 18]
```diff
@@ -31,84 +34,126 @@ class TennisTVIE(InfoExtractor):
...
+            self.REFRESH_TOKEN = cookies.get('refresh_token').value
+            return
+
+        else:
+            self.report_warning('Site requires a subscribed account. Either pass in the cookies or login credentials.')
+
+    def get_token(self, video_id, payload):
+        res = self._download_json('https://sso.tennistv.com/auth/realms/TennisTV/protocol/openid-connect/token', video_id,
+                                  'Fetching tokens', 'Unable to fetch tokens', headers=self.headers,
+                                  data=urlencode_postdata(payload))
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:107]:
```suggestion
        res = self._download_json('https://sso.tennistv.com/auth/realms/TennisTV/protocol/openid-connect/token',
                                  video_id, 'Fetching tokens', 'Unable to fetch tokens',
                                  headers=self.headers, data=urlencode_postdata(payload))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by Bricio** [comment]
Old code returns city, season and round.
image

https://github.com/yt-dlp/yt-dlp/blob/e7870111e83033e0ac728d5a2d565d1eb146c335/yt_dlp/extractor/tennistv.py#L110-L112

You can still collect that info, in attribute values of the same div that you are collecting description now.
image

It will be good to return same data.

**Review by pukkandan — CHANGES_REQUESTED** [review]
partial review

File: yt_dlp/extractor/tennistv.py [hunk 19]
```diff
@@ -1,13 +1,15 @@
 # coding: utf-8
 from __future__ import unicode_literals
 
-import json
-
 from .common import InfoExtractor
 
+import urllib.parse
+
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:7]:
```suggestion
import urllib.parse

from .common import InfoExtractor
```

File: yt_dlp/extractor/tennistv.py [hunk 20]
```diff
@@ -1,13 +1,15 @@
...
 from ..utils import (
     ExtractorError,
     unified_timestamp,
+    urlencode_postdata,
+    random_uuidv4,
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:12]:
```suggestion
    random_uuidv4,
    unified_timestamp,
    urlencode_postdata,
```

File: yt_dlp/extractor/tennistv.py [hunk 21]
```diff
@@ -31,84 +33,126 @@ class TennisTVIE(InfoExtractor):
...
 
-        self._session_token = login_result['sessionToken']
+        login_page = self._download_webpage(BASE_URL, None, query={
+            'client_id': 'tennis-tv-web',
+            'redirect_uri': 'https://tennistv.com',
+            'response_mode': 'fragment',
+            'response_type': 'code',
+            'scope': 'openid'
+        })
+        post_url = self._html_search_regex(r'action=["\'](.+?)["\']\s+?method=["\']post["\']', login_page, None)
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:62]:
```suggestion
        post_url = self._html_search_regex(r'action=["\']([^"\']+?)["\']\s+method=["\']post["\']', login_page, None)
```

File: yt_dlp/extractor/tennistv.py [hunk 22]
```diff
@@ -31,84 +33,126 @@ class TennisTVIE(InfoExtractor):
...
-            raise ExtractorError('No login info available, needed for using %s.' % self.IE_NAME, expected=True)
+    ACCESS_TOKEN = None
+    REFRESH_TOKEN = None
+    PARTNER_ID = 3001482
+    _FORMAT_URL = 'https://open.http.mp.streamamg.com/p/{partner}/sp/{partner}00/playManifest/entryId/{entry}/format/applehttp/protocol/https/a.m3u8?ks={session}'
+    headers = {
+        'origin': 'https://www.tennistv.com',
+        'referer': 'https://www.tennistv.com/',
+        'content-Type': 'application/x-www-form-urlencoded'
+    }
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:44]:
```suggestion
    access_token, refresh_token = None, None
    _FORMAT_URL = 'https://open.http.mp.streamamg.com/p/{partner}/sp/{partner}00/playManifest/entryId/{entry}/format/applehttp/protocol/https/a.m3u8?ks={session}'
    HEADERS = {
        'origin': 'https://www.tennistv.com',
        'referer': 'https://www.tennistv.com/',
        'content-Type': 'application/x-www-form-urlencoded'
    }
```
normally, caps are used for constants

File: yt_dlp/extractor/tennistv.py [hunk 23]
```diff
@@ -31,84 +33,126 @@ class TennisTVIE(InfoExtractor):
...
+            'state': random_uuidv4(),
+            'response_mode': 'fragment',
+            'response_type': 'code',
+            'scope': 'openid',
+            'nonce': random_uuidv4(),
+            'prompt': 'none'
+        })
+
+        self.get_token(None, {
+            'code': urllib.parse.parse_qs(handle.geturl()).get('code')[-1],
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:81]:
```suggestion
            'code': urllib.parse.parse_qs(handle.geturl())['code'][-1],
```

File: yt_dlp/extractor/tennistv.py [hunk 24]
```diff
@@ -31,84 +33,126 @@ class TennisTVIE(InfoExtractor):
...
+    _FORMAT_URL = 'https://open.http.mp.streamamg.com/p/{partner}/sp/{partner}00/playManifest/entryId/{entry}/format/applehttp/protocol/https/a.m3u8?ks={session}'
+    headers = {
+        'origin': 'https://www.tennistv.com',
+        'referer': 'https://www.tennistv.com/',
+        'content-Type': 'application/x-www-form-urlencoded'
+    }
+
+    def _perform_login(self, username, password):
 
+        BASE_URL = 'https://sso.tennistv.com/auth/realms/TennisTV/protocol/openid-connect/auth'
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:48]: makes more sense to put `https://sso.tennistv.com/auth/realms/TennisTV/protocol` in a class var instead imo. This is also used in `get_token`

File: yt_dlp/extractor/tennistv.py [hunk 25]
```diff
@@ -31,84 +33,126 @@ class TennisTVIE(InfoExtractor):
...
+            self._FORMAT_URL.format(partner=self.PARTNER_ID, entry=entryid, session=k_session), video_id)
+
+        self._sort_formats(formats)
 
         return {
             'id': video_id,
-            'title': title,
-            'description': description,
+            'title': self._html_search_regex((r'<title>(.+?)</title>', *self._og_regexes('title')),
+                webpage, 'Title', fatal=False),
```
Comment (pukkandan) [inline yt_dlp/extractor/tennistv.py:146]:
```suggestion
            'title': self._html_extract_title(webpage) or self._og_search_title(webpage),
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
and merge with master

**Pull request merged**